### ISSUE TITLE
Specialized Trainer Abstractions and RuntimeConfig for the Kubeflow SDK

### ISSUE THREAD

**Issue by szaher** [kubeflow/sdk#285]
## Problem

The current SDK offers two trainer abstractions that leave a usability gap for the majority of distributed training workloads:

- **`CustomTrainer`** is generic but requires users to manually look up runtimes by name, provides no framework-specific validation, and mixes runtime-environment settings (`packages_to_install`, `pip_index_urls`, `env`) with scaling/training concerns (`num_nodes`, `resources_per_node`, `func`) in a single flat dataclass.

- **`BuiltinTrainer`** is narrow, hardcoded to `TorchTuneConfig`, and cannot be extended to other config-driven frameworks without modifying the class itself.

For standard use cases like "run this PyTorch DDP function on N nodes" or "run this MPI script across a cluster," neither abstraction fits cleanly.

## Proposal

1. **Specialized, framework-aware trainers** (`TorchTrainer`, `MPITrainer`, `JAXTrainer`, `XGBoostTrainer`) that:
   - Auto-discover the correct `ClusterTrainingRuntime` using the existing `trainer.kubeflow.org/framework` label
   - Validate trainer/runtime compatibility at submission time (not at execution time)
   - Expose typed, framework-specific arguments (e.g., `max_restarts` for Torch, `num_proc_per_node` for MPI)
   - Share a common `BaseTrainer` abstract interface enabling community extensions

2. **`RuntimeConfig` dataclass** that cleanly separates per-job runtime environment settings (packages, pip config, env vars) from training logic, passed as a dedicated parameter to `TrainerClient.train()`.

Both changes are additive and 100% backward-compatible. Existing `CustomTrainer` and `BuiltinTrainer` usage is unaffected.

## Example

**Before:**
```python
runtime = client.get_runtime("torch-distributed")
job_id = client.train(
    runtime=runtime,
    trainer=CustomTrainer(
        func=train_fn,
        func_args={"lr": 1e-4},
        packages_to_install=["transformers"],
        env={"NCCL_DEBUG": "INFO"},
        num_nodes=4,
        resources_per_node={"gpu": 1},
    ),
)
```

**After:**
```python
job_id = client.train(
    trainer=TorchTrainer(
        func=train_fn,
        func_args={"lr": 1e-4},
        num_nodes=4,
        resources_per_node={"gpu": 1},
        max_restarts=3,
    ),
    runtime_config=RuntimeConfig(
        packages=["transformers"],
        env={"NCCL_DEBUG": "INFO"},
    ),
)
```

## References

- KEP-2170: [Kubeflow Trainer V2 API](https://github.com/kubeflow/trainer/blob/master/docs/proposals/2170-kubeflow-trainer-v2/README.md)
- Runtime framework label: [Runtime Guide](https://www.kubeflow.org/docs/components/trainer/operator-guides/runtime/)
- Full design proposal will be submitted as a PR to this repo

/cc @kubeflow/kubeflow-sdk-team

**Comment by preetbiswas12** [comment]
Hi @szaher

I would like to work on this issue.

I think it makes sense to separate runtime concerns using a specific RuntimeConfig and introduce framework-aware trainers (like TorchTrainer and MPITrainer), especially from the standpoint of API design. I concur that separating environment configuration from scaling logic and verifying trainer/runtime compatibility at submission time would greatly enhance usability for typical distributed workloads.I would be pleased to begin drafting the BaseTrainer abstraction and a TorchTrainer implementation as a first step if this strategy fits with the current project.

Tell me if I can be assigned to this problem, please.

**Comment by krishdef7** [comment]
This proposal looks great, especially the separation between framework-aware trainers and RuntimeConfig.

One thing I’m particularly interested in is how the BaseTrainer abstraction would support Tier-2 trainers built on top of existing frameworks such as Hugging Face Transformers or TRL-based LLM workflows.

For example, a HuggingFaceTrainer could wrap the Transformers Trainer API while targeting a Torch runtime underneath, but expose higher-level configuration objects (training args, datasets, PEFT configs, etc.).

Do you envision these Tier-2 trainers living inside the Kubeflow SDK itself, or as external extensions built on top of the BaseTrainer interface?

I’ve been exploring LLM training workflows on Kubeflow Trainer and would be very interested in prototyping something along these lines once the BaseTrainer abstraction lands.
